site/content/posts/2022-01-18-idp-refresh-tls-ciphers-for-compliance.md

```diff
@@ -0,0 +1,131 @@
+
+Example changes to Okta OIDCProvider CRD:
+
+```yaml
+apiVersion: idp.supervisor.pinniped.dev/v1alpha1
+kind: OIDCIdentityProvider
+metadata:
+  namespace: pinniped-supervisor
+  name: okta
+spec:
+
+  authorizationConfig:
+
+    # Request any scopes other than "openid" for claims besides
+    # the default claims in your token. The "openid" scope is always
+    # included.
+    #
+    # To learn more about how to customize the claims returned, see here:
+    # https://developer.okta.com/docs/guides/customize-tokens-returned-from-okta/overview/
+    additionalScopes: [offline_access, groups, email]
+
+    # If you would also like to allow your end users to authenticate using
+    # a password grant, then change this to true. Password grants only work
+    # with applications created in Okta as "Native Applications".
+    allowPasswordGrant: false
+```
+
+Refer complete yaml format at [how to configure Okta as IDP with Supervisor]({{< ref "docs/howto/configure-supervisor-with-okta.md" >}})
+
+Inside Okta, when you create the Application, make sure to select refresh tokens as the Grant type along with Authorization code. See below:
+
+![Okta screenshot with Grant types](/docs/img/refresh-token-grant-okta.png)
+
+####  When your OIDC IDP cannot return refresh tokens
+
+In the case where your IDP is not capable of returning refresh tokens, for example if you are using Dex with the SAML connector, Pinniped will refresh the session using **Access tokens**. Pinniped will validate the Access tokens against the **userinfo endpoint** of your IDP. You are required to provide userinfo endpoint or refresh tokens for session validation. Your login will fail if neither of the two options is provided.  
+
+If your access tokens have a lifetime shorter than 3 hours, Pinniped will issue a **warning** that gets displayed to the end user’s CLI notifying them that the access token TTL is less than 3 hours and the end user will need to re-login again after it expires. Here, the administrator has the option to increase the access token's expiration time in their upstream IDP.  A more detailed, admin focused warning is also emitted in the Pinniped supervisor pod logs.
+
+### What about LDAP / Active Directory IDP changes?
+
+LDAP does not have a concept of sessions or refresh tokens. Hence we run LDAP queries against the LDAP or AD IDP to simulate the refresh. For LDAP, we validate if the LDAP entry exists with no changes to Pinniped *UID and Username* fields. For AD, we validate all the LDAP checks mentioned earlier and we also validate the user's password and account status if locked or disabled. If any of the LDAP or AD checks fail, the user will be able to use the existing Pinniped session until it expires in about 5 minutes.
```

Comment:
```suggestion
LDAP does not have a concept of sessions or refresh tokens. Hence we run LDAP queries against the LDAP or AD IDP to approximate a refresh. For LDAP, we validate if the LDAP entry still exists with no changes to Pinniped UID and username fields. For AD, we validate the same LDAP checks and we also validate the user's password has not changed since the original login and their account is not locked or disabled.
```